### DISCUSSION TITLE
How to do federated learning using onnxruntime?

### DISCUSSION THREAD

**Discussion by KartikSgh** [microsoft/onnxruntime#19490] [Other Q&A] [answered]
In the latest release, v1.17.0, under the Training section, for on-device training, it is written - "**On-Device training will now support training on the web. This release focuses on federated learning and developer exploration scenarios. More features coming soon in future releases.**".

My questions:
- How to use this feature to do federated learning on edge devices. There is no documentation that provides the steps to do it. 
- Is there any open-source framework that facilitates federated learning using onnx model?

**Comment by carzh** [comment, marked as answer]
Hi, [this](https://cloudblogs.microsoft.com/opensource/2024/02/06/on-device-training-training-a-model-in-browser/) blog post goes into a bit more detail.

onnxruntime-web 1.17.0 package now includes training APIs that will perform the edge device training portion of federated learning. The [APIs ](https://onnxruntime.ai/docs/api/js/interfaces/TrainingSession-1.html) loadParametersBuffer and getContiguousParameters enable federated learning -- the buffer of parameters should be sent from the edge device to the central server with the federated learning framework. After the central server performs weight updates on the central model, another buffer of parameters is sent to edge devices.
